GVFS/GVFS.Common/Database/GVFSDatabase.cs

```diff
@@ -0,0 +1,135 @@
+
+            string folderPath = Path.GetDirectoryName(this.databasePath);
+            fileSystem.CreateDirectory(folderPath);
+
+            for (int i = 0; i < InitialPooledConnections; i++)
+            {
+                this.connectionPool.Add(this.OpenNewConnection());
+            }
+
+            this.Initialize();
+            this.CreateTables();
+        }
+
+        public interface IPooledConnection : IDisposable
+        {
+            SqliteConnection Connection { get; }
+        }
+
+        public void Dispose()
+        {
+            this.disposed = true;
+            this.connectionPool.CompleteAdding();
+            while (!this.connectionPool.IsCompleted && this.connectionPool.TryTake(out SqliteConnection connection))
+            {
+                connection.Dispose();
+            }
+        }
+
+        public IPooledConnection GetPooledConnection()
+        {
+            SqliteConnection connection;
+            if (!this.connectionPool.TryTake(out connection, millisecondsTimeout: MillisecondsWaitingToGetConnection))
+            {
+                connection = this.OpenNewConnection();
+            }
+
+            return new GVFSConnection(this, connection);
+        }
+
+        private void Initialize()
+        {
+            using (IPooledConnection pooled = this.GetPooledConnection())
+            using (SqliteCommand command = pooled.Connection.CreateCommand())
+            {
+                command.CommandText = $"PRAGMA journal_mode=WAL;";
+                command.ExecuteNonQuery();
+                command.CommandText = $"PRAGMA cache_size=-40000;";
+                command.ExecuteNonQuery();
+                command.CommandText = $"PRAGMA synchronous=NORMAL;";
+                command.ExecuteNonQuery();
+                command.CommandText = $"PRAGMA user_version;";
+                object userVersion = command.ExecuteScalar();
+                if (userVersion == null || Convert.ToInt64(userVersion) < 1)
+                {
+                    command.CommandText = $"PRAGMA user_version=1;";
+                    command.ExecuteNonQuery();
+                }
+            }
+        }
+
+        private void CreateTables()
+        {
+            using (IPooledConnection pooled = this.GetPooledConnection())
+            using (SqliteCommand command = pooled.Connection.CreateCommand())
+            {
+                Placeholders.CreateTable(command);
+            }
+        }
+
+        private SqliteConnection OpenNewConnection()
+        {
+            SqliteConnection connection = new SqliteConnection(this.sqliteConnectionString);
+            connection.Open();
+            return connection;
+        }
+
+        private void ReturnToPool(SqliteConnection connection)
+        {
+            if (this.disposed)
+            {
+                connection?.Dispose();
+            }
+            else if (!this.connectionPool.TryAdd(connection))

```

Comment:
Can this throw? For example, if the pool has been marked complete with regards to additions?

https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.blockingcollection-1.tryadd?view=netframework-4.8

Reply:
I set the `disposed` to `true` before calling `CompleteAdding` but I see there is still a race if this code checked `disposed` and then the other code ran that sets the `disposed` and then calls `CompleteAdding` before this code runs the `TryAdd`.   I will have to handle this.  Thanks!